### PR TITLE
Show container volumes in the dashboard

### LINKED ISSUE #3969: Make container volumes visible from the dashboard
Volume information doesn't show up anywhere in the dashboard. It should.

### PR DESCRIPTION
**Pull request by drewnoakes**
## Description

Fixes #3969

This plumbs through information about volumes mounted to containers, for display in the dashboard alongside other container details.

image

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [x] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5572)

### CODE REVIEW

File: src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs [hunk 1]
```diff
@@ -77,6 +82,14 @@ public sealed record EnvironmentVariableSnapshot(string Name, string? Value, boo
 /// <param name="IsInternal">Determines if this url is internal.</param>
 public sealed record UrlSnapshot(string Name, string Url, bool IsInternal);
 
+/// <summary>
+/// A snapshot of a volume, mounted to a container.
+/// </summary>
+/// <param name="Source">The name of the volume. Can be <c>null</c> if the mount is an anonymous volume.</param>
+/// <param name="Target">The target of the mount.</param>
+/// <param name="IsReadOnly">Whether the volume mount is read-only or not.</param>
+public sealed record VolumeSnapshot(string? Source, string Target, bool IsReadOnly);
```
Comment (mitchdenny) [inline src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:91]: I'm wondering whether we should capture the volume type here. You should be able to tell intuitively from the source but might be worth making it obvious whether it is a bind mount or a volume mount.

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor [hunk 2]
```diff
@@ -121,6 +121,22 @@
                               HighlightText="@_filter"
                               GridTemplateColumns="1fr 1.5fr" />
             </FluentAccordionItem>
+            <FluentAccordionItem Heading="@ControlStringsLoc[nameof(ControlsStrings.ResourceDetailsVolumesHeader)]" Expanded="true">
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor:124]: This is only valid for containers, right? I think there should be an if around this section to only show for containers.

File: src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs (hunk 1 above)
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:91]: I know the pattern here is to use records, but I have concerns about using records in public API. Let's say we add new fields in the future, how do we do that with a record?

I think a normal C# type with init properties would be easier to expand in the future.

File: src/Aspire.Hosting/Dashboard/proto/resource_service.proto [hunk 3]
```diff
@@ -117,6 +117,12 @@ message Url {
     bool is_internal = 3;
 }
 
+message Volume {
+    string source = 1;
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dashboard/proto/resource_service.proto:121]: Add comments to new proto fields.

File: src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:91]: I'll add it to the model. In the UI we currently use name/value pairs in the details view, so won't show it (or `IsReadOnly`) without further work.

Comment (drewnoakes) [reply src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs]: That's fair, but let's do that separately and keep the code consistent here.

Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs]: Agreed

**Comment by davidfowl** [comment]
@drewnoakes Seems like there's a conflict with @mitchdenny's PR. 

Can you show an example of a bind mount?

This change looks good! Thanks!

File: src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs]: I found a way to show it: 

image

The visualizer/copy button works with the raw value only (`/var/lib/rabbitmq`), which I think makes the most sense.

This shows both the mount type, and whether the mount is readonly or not. In this case I hard-coded the value to true, as I didn't see an example in the playground.

**Comment by drewnoakes** [comment]
Conflicts fixed.

Found a way to visualize mount type and read only state:

image

The visualizer/copy button works with the raw value only (`/var/lib/rabbitmq`), which I think makes the most sense.

This shows both the mount type, and whether the mount is read only or not. In this case I hard-coded the value to true, as I didn't see an example in the playground.

**Comment by davidfowl** [comment]
Bind mounts have a physical path for the source

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor [hunk 4]
```diff
@@ -121,6 +121,35 @@
...
+                                  IsNameSortable="false"
+                                  IsValueSortable="false"
+                                  EnableValueMasking="false"
+                                  HighlightText="@_filter"
+                                  GridTemplateColumns="1fr 1.5fr">
+                        <ExtraValueContent>
+                            <div>
+                                @if (context.IsReadOnly)
+                                {
+                                    <FluentIcon Value="@(new Icons.Regular.Size12.LockClosed())" />
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor:145]: The icon is a little small. We use 16x16 icons with inline text in other places. For example, the state icon on the resources screen:

https://github.com/dotnet/aspire/blob/b6e3e16393835fd67ca482da440ac7efb158f8c1/src/Aspire.Dashboard/Components/ResourcesGridColumns/StateColumnDisplay.razor#L14-L17

Can you make this 16x16 and make the icon and text vertically middle aligned with each other?

**Pull request merged**